forgerock-auth/src/main/java/org/forgerock/android/auth/webauthn/WebAuthnAuthentication.kt

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2022 ForgeRock. All rights reserved.
+ *
+ *  This software may be modified and distributed under the terms
+ *  of the MIT license. See the LICENSE file for details.
+ */
+package org.forgerock.android.auth.webauthn
+
+import android.content.Context
+import android.util.Base64
+import com.google.android.gms.fido.Fido
+import com.google.android.gms.fido.fido2.api.common.AuthenticatorAssertionResponse
+import com.google.android.gms.fido.fido2.api.common.PublicKeyCredential
+import com.google.android.gms.fido.fido2.api.common.PublicKeyCredentialDescriptor
+import com.google.android.gms.fido.fido2.api.common.PublicKeyCredentialRequestOptions
+import kotlinx.coroutines.tasks.await
+import org.forgerock.android.auth.WebAuthnDataRepository
+import org.json.JSONArray
+import org.json.JSONException
+import org.json.JSONObject
+
+/**
+ * Handle WebAuthn Authentication
+ */
+open class WebAuthnAuthentication() : WebAuthn() {
+
+    internal lateinit var options: PublicKeyCredentialRequestOptions
+
+    constructor(input: JSONObject) : this() {
+        val challenge = Base64.decode(input.getString(CHALLENGE), Base64.NO_WRAP)
+        val relayingPartyId = getRelyingPartyId(input)
+        val timeout = input.optString(TIMEOUT, TIMEOUT_DEFAULT).toDouble() / 1000
+        val allowCredentials = getAllowCredentials(input)
+        //val userVerification = input.optString(USER_VERIFICATION, null)
+        options = PublicKeyCredentialRequestOptions.Builder()
+            .setAllowList(allowCredentials)
+            .setRpId(relayingPartyId)
+            .setChallenge(challenge)
+            .setTimeoutSeconds(timeout)
+            .build()
+    }
+
+    constructor(options: PublicKeyCredentialRequestOptions) : this() {
+        this.options = options
+    }
+
+    /**
+     * Parse and retrieve all the allow credentials
+     *
+     * @param value The json from WebAuthn Authentication Node
+     * @return The parsed PublicKeyCredentialDescriptor
+     * @throws JSONException Failed to parse the Json
+     */
+    @Throws(JSONException::class)
+    protected fun getAllowCredentials(value: JSONObject): List<PublicKeyCredentialDescriptor> {
+        var allowCredentials: JSONArray? = JSONArray()
+        if (value.has(_ALLOW_CREDENTIALS)) {
+            allowCredentials = value.getJSONArray(_ALLOW_CREDENTIALS)
+        } else if (value.has(ALLOW_CREDENTIALS)) {
+            val allowCredentialString = value.getString(ALLOW_CREDENTIALS)
+                .replace("(allowCredentials: |new Int8Array\\(|\\).buffer )".toRegex(), "")
+            if (allowCredentialString.trim().isNotEmpty()) {
+                allowCredentials = JSONArray(allowCredentialString)
+            }
+        }
+        return getCredentials(allowCredentials!!)
```

Comment:
var allowCredentials: JSONArray? = JSONArray() - is this need to be optional . if yes may be we can do getCredentials(allowCredentials ?? JSONArray())

Reply:
It is not optional, updated.